### DISCUSSION TITLE
Using x-for in Alpine.data causes error on 3.4.2

### DISCUSSION THREAD

**Discussion by mimamuh** [alpinejs/alpine#2266] [7. Bugs (archived)]
Hi, I get the following error using `x-for` within `Alpine.data(...)` like so:

```html
<div x-data="projectsList">
    <template x-bind="projectLoop">
          ...
    </template>
</div>
<script>
    document.addEventListener('alpine:init', () => {
        Alpine.data('projectsList', ({ projects = [...] }) => ({
            projects,

            projectLoop: {
                ['x-for']() {
                    return '(project, index) in projects';
                },
            },
        }))
    })
</script>
```

##### Error:
```
Alpine Expression Error: expression.match is not a function
Expression: "projectLoop"
<template x-bind=​"projectLoop">​…​</template>​

module.esm.js?dd52:3109 Uncaught TypeError: expression.match is not a function
    at parseForExpression (VM56177 module.esm.js:3109)
    at Function.eval (VM56177 module.esm.js:2998)
    at fullHandler (VM56177 module.esm.js:1771)
    at eval (VM56177 module.esm.js:2931)
    at Array.map (<anonymous>)
    at eval (VM56177 module.esm.js:2929)
    at runIfTypeOfFunction (VM56177 module.esm.js:1691)
    at eval (VM56177 module.esm.js:1674)
    at tryCatch (VM56177 module.esm.js:1696)
    at eval (VM56177 module.esm.js:2921)
```

##### What works fine is:

1) It works fine when I use `x-for="(project, index) in projects"` directly on the `<template x-for="(project, index) in projects">` but not using it through `Alpine.data(...)`.

2) It works fine when I use a property instead of a method like so:
```JS
            projectLoop: {
                // ['x-for']() { return '(project, index) in projects'; }
                ['x-for']: '(project, index) in projects', 
            },
```

BUT: Based on the [docs here](https://alpinejs.dev/directives/bind#bind-directives) it should be actually like I did it in the erroneous example above: Using a method for it.

##### Question
What is the right way to do it? In my understanding both should work: Using a property or a method should work here. But only the property thing is working so far for me.

**Comment by SimoTod** [comment]
If you use a function, it must return the result of the evaluated function not a string to evaluate. If you look at the example with x-show, it returns the boolean not the string representing the boolean. Basically if you use a function, you skip the eval stage which converts a string to javascript code. I hope it makes sense.

**Reply by mimamuh** [reply]
Thx. Yeah, that makes sense. That's how I did it before with simple scenarios like `x-show`.
But in the context of `x-for`, how would it be done with a function? What's expected to be returned here? Can you make an example?

**Reply by mimamuh** [reply]
When I think about it what you said, probably `['x-for']: '(project, index) in projects'` is the only way how to do it for `x-for` here, right?

**Reply by mimamuh** [reply]
The only thing which confuses me then are the [docs here](https://alpinejs.dev/directives/bind#bind-directives), claiming:

> When the directive being "bound" or "applied" is x-for, you should return a normal expression string from the callback. For example: ['x-for']() { return 'item in items' }

But that doesn't work for me and – when I understood it right – contradicts what you said.

**Reply by SimoTod** [reply]
I wasn't aware of that part of the docs. If there is meant to be a special exception for x-for, then it's a bug
